vg/vg_test.go

```diff
@@ -0,0 +1,129 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package vg_test
+
+import (
+	"bytes"
+	"flag"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"os"
+	"path/filepath"
+	"testing"
+
+	"github.com/gonum/plot"
+	"github.com/gonum/plot/plotter"
+	"github.com/gonum/plot/vg"
+	"github.com/gonum/plot/vg/draw"
+	"github.com/gonum/plot/vg/vgeps"
+	"github.com/gonum/plot/vg/vgimg"
+	"github.com/gonum/plot/vg/vgpdf"
+	"github.com/gonum/plot/vg/vgsvg"
+)
+
+var generateTestData = flag.Bool("good", false, "Specifies that the current state is good.")
+
+// TestLineWidth tests output against test images generated by
+// running tests with -tag good.
+func TestLineWidth(t *testing.T) {
+	formats := []string{
+		// TODO: Add logic to cope with run to run eps differences.
+		"pdf",
+		"svg",
+		"png",
+		"tiff",
+		"jpg",
+	}
+
+	const (
+		width  = 100
+		height = 100
+	)
+
+	for _, w := range []vg.Length{-1, 0, 1} {
+		for _, typ := range formats {
+			p, err := lines(w)
+			if err != nil {
+				log.Fatalf("failed to create plot for %v:%s: %v", w, typ, err)
+			}
+
+			var c interface {
+				vg.Canvas
+				Size() (w, h vg.Length)
+				io.WriterTo
+			}
+
+			name := filepath.Join(".", "testdata", fmt.Sprintf("width_%v.%s", w, typ))
+			switch typ {
+			case "eps":
+				c = vgeps.NewTitle(width, height, name)
+
+			case "jpg":
+				c = vgimg.JpegCanvas{Canvas: vgimg.New(width, height)}
+
+			case "pdf":
+				c = vgpdf.New(width, height)
+
+			case "png":
+				c = vgimg.PngCanvas{Canvas: vgimg.New(width, height)}
+
+			case "svg":
+				c = vgsvg.New(width, height)
+
+			case "tiff":
+				c = vgimg.TiffCanvas{Canvas: vgimg.New(width, height)}
+
+			default:
+				panic("bad typ string")
+			}
+
+			p.Draw(draw.New(c))
+
+			var buf bytes.Buffer
+			if _, err = c.WriteTo(&buf); err != nil {
+				t.Errorf("failed to write plot for %v:%s: %v", w, typ, err)
+			}
+
+			// Recreate Golden images.
+			if *generateTestData {
+				err = p.Save(width, height, name)
+				if err != nil {
+					log.Fatalf("failed to save %q: %v", name, err)
+				}
+			}
+
+			f, err := os.Open(name)
+			if err != nil {
+				t.Fatalf("failed to open test image: %v", err)
+			}
```

Comment:
It's fine to defer close in a loop and it keeps with the idiom, but closing
after the read is just great. The point is: don't forget to close.

On Sat, Mar 28, 2015, 7:24 AM Dan Kortschak <notifications@github.com>
wrote:

> In vg/vg_test.go
> <https://github.com/gonum/plot/pull/186#discussion_r27344329>:
>
> > +			if _, err = c.WriteTo(&buf); err != nil {
> > +				t.Errorf("failed to write plot for %v:%s: %v", w, typ, err)
> > +			}
> > +
> > +			// Recreate Golden images.
> > +			if *generateTestData {
> > +				err = p.Save(width, height, name)
> > +				if err != nil {
> > +					log.Fatalf("failed to save %q: %v", name, err)
> > +				}
> > +			}
> > +
> > +			f, err := os.Open(name)
> > +			if err != nil {
> > +				t.Fatalf("failed to open test image: %v", err)
> > +			}
>
> defered close is no good here - in a loop and early exits are exits, not
> returns.
>
> I'll add f.Close after the ReadAll call.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/gonum/plot/pull/186/files#r27344329>.
>